numpy/_core/src/multiarray/argfunc.dispatch.cpp

```diff
@@ -0,0 +1,522 @@
+#include "simd/simd.h"
+#include "numpy/npy_math.h"
+#include "numpy/npy_common.h"
+#include "common.hpp"
+#include "arraytypes.h"
+#include "simd/simd.hpp"
+#include <hwy/highway.h>
+
+#define MIN(a,b) (((a)<(b))?(a):(b))
+
+namespace {
+using namespace np::simd;
+
+template <typename T>
+struct OpGt {
+#if NPY_HWY
+    template <typename V, typename = std::enable_if_t<kSupportLane<T>>>
+    HWY_INLINE HWY_ATTR auto operator()(const V &a, const V &b) const { 
+        return hn::Gt(a, b);
+    }
+#endif
+    HWY_INLINE bool operator()(T a, T b) {
+        return a > b; 
+    }
+
+    HWY_INLINE bool negated_op(T a, T b) {
+        return a <= b;
+    }
+};
+
+template <>
+struct OpGt<long double> {
+    HWY_INLINE bool operator()(long double a, long double b) {
+        return a > b; 
+    }
+
+    HWY_INLINE bool negated_op(long double a, long double b) {
+        return a <= b;
+    }
+};
+
+template <typename T>
+struct OpLt {
+#if NPY_HWY
+    template <typename V, typename = std::enable_if_t<kSupportLane<T>>>
+    HWY_INLINE HWY_ATTR auto operator()(const V &a, const V &b) const { 
+        return hn::Lt(a, b); 
+    }
+#endif
+    HWY_INLINE bool operator()(T a, T b) { 
+        return a < b; 
+    }
+
+    HWY_INLINE bool negated_op(T a, T b) {
+        return a >= b;
+    }
+};
+
+template <>
+struct OpLt<long double> {
+    HWY_INLINE bool operator()(long double a, long double b) {
+        return a < b; 
+    }
+
+    HWY_INLINE bool negated_op(long double a, long double b) {
+        return a >= b;
+    }
+};
+
+#if NPY_HWY
+template <typename T, typename Op>
+static HWY_INLINE HWY_ATTR npy_intp
+simd_argfunc_small(T *ip, npy_intp len)
+{
+    //static_assert(kMaxLanes<uint8_t> <= 64,
+    //    "the following 8/16-bit argmax kernel isn't applicable for larger SIMD");
+    /* TODO: add special loop for large SIMD width.
+       i.e avoid unroll by x4 should be numerically safe till 2048-bit SIMD width
+       or maybe expand the indices to 32|64-bit vectors(slower).  */
+
+    using UnsignedT = std::conditional_t<sizeof(T) == 1, uint8_t, uint16_t>;
+    constexpr npy_intp idx_max = (sizeof(T) == 1) ? NPY_MAX_UINT8 : NPY_MAX_UINT16;
+
+    Op op_func;
+    T s_acc = *ip;
+    npy_intp ret_idx = 0, i = 0;
+
+    HWY_LANES_CONSTEXPR int vstep = Lanes<T>();
+    const int wstep = vstep*4;
+    const auto vindices_0 = hn::Iota(_Tag<UnsignedT>(), UnsignedT(0));
```

Comment:
```suggestion
    const Vec<U> vindices_0 = Iota(U{0});
```
Expand the wrapper and avoid `auto` for type safety.